lib/pavilion/schedulers/plugins/slurm.py

```diff
@@ -313,72 +389,42 @@ def _get_config_elems(self):
-                        "Invalid Node List: '{}'. Syntax error in item '{}'. "
-                        "Node lists components be a hostname or hostname "
-                        "prefix followed by a range of node numbers. "
-                        "Ex: foo003,foo0[10-20],foo[103-104],foo[10,12-14],foo-m11-16"
-                        .format(node_list, part)
-                    )
-
-            # If all the parts matched, then it's an overall format issue.
-            raise ValueError("Invalid Node List: '{}' "
-                             "Good Example: foo003,foo0[10-20],"
-                             "foo[103-104], foo[10,12-14]")
-
-        nodes = []
-        prev = ""
-        for part in node_list.split(','):
-            if prev:
-                part = prev + "," + part
-                prev = ""
-            if '[' in part and ']' not in part:
-                prev = part
-                continue
-            match = cls.NODE_BRACKET_FORMAT_RE.match(part)
-            if match:
-                host, nodelist = match.groups()
-                for node in nodelist.split(","):
-                    if '-' in node:
-                        start, end = node.split('-')
-                        digits = min(len(start), len(end))
-                        if int(end) < int(start):
-                            raise ValueError(
-                                "In node list '{}' part '{}', node range ends "
-                                "before it starts."
-                                .format(node_list, part)
-                            )
-                        for i in range(int(start), int(end)+1):
-                            node = ('{base}{num:0{digits}d}'
-                                    .format(base=host, num=i, digits=digits))
-                            nodes.append(node)
-                    else:
-                        node = ('{base}{num}'
-                                .format(base=host, num=node))
-                        nodes.append(node)
+        # Split the nodelist on commas to seperate sections.
+        node_sections = node_list.split(',')
+        allnodes = []
+        inner_split = False
+        inner_collect = []
+
+        # Collect sections so that each element is either a single node string
+        # or a node set descriptor.
+        for node_section in node_sections:
+            if inner_split:
+                inner_collect.append(node_section)
+                if "]" in node_section:
+                    inner_split = False
+                    allnodes.append(','.join(inner_collect))
+                    inner_collect = []
+            elif "[" not in node_section:
+                allnodes.append(node_section)
             else:
-                nodes.append(part)
+                if "]" in node_section:
+                    allnodes.append(node_section)
+                else:
+                    inner_collect.append(node_section)
+                    inner_split = True
```

Comment:
This is taking the Python way and trying to fix it. It's much cleaner and more efficient to do this the C way:

```python
 start = 0
depth = 0
node_list = []
for i, char in enumerate(test_data):
    if char == '[':
        if depth == 0:
            depth = 1
        else:
            raise ValueError("Extra open bracket in '{}'".format(test_data[start:i+1]))
    if char == ']':
        if depth != 1:
            raise ValueError("Extra close bracket in '{}'".format(test_data[start:i+1]))
        else:
            depth = 0
    if char == ',' and depth == 0:
        node_list.extend(parse_node_list_element(test_data[start:i])
        start = i+1 
```